timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,954 @@
+        if len(nbs) == 4:
+            chiral_atoms.add(a_idx)
+        elif len(nbs) == 3 and mol_geom[a_idx] == LocalGeometry.G3_PYRAMIDAL:
+            # if in ring, or is sulfur or phosphorus, this may not be guaranteed by
+            if a.GetAtomicNum() == 16 or a.GetAtomicNum() == 15:
+                assert 0, "Pyramidal Phosphorous or Sulfur detected."
+        elif len(nbs) > 4:
+            assert 0, "More than four neighbors found"
+    return chiral_atoms
+
+
+def check_bond_stability(j, k, bond_idxs, bond_params):
+    # tbd this should really be based on the pdf, we should expect to see
+    # near zero probability at bond lengths < 0.5 Angstroms
+    canonical_bonds = dict()
+    for idxs, params in zip(bond_idxs, bond_params):
+        canonical_bonds[dummy.canonicalize_bond(idxs)] = params
+
+    jk = dummy.canonicalize_bond((j, k))
+
+    if jk not in canonical_bonds:
+        return False
+    elif canonical_bonds[jk][0] < 50.0:
+        return False
+    elif canonical_bonds[jk][1] < 0.05:
+        return False
+
+    return True
+
+
+def check_angle_stability(j, k, l, angle_idxs, angle_params):
+
+    canonical_angles = dict()
+    for idxs, params in zip(angle_idxs, angle_params):
+        force_const, theta = params
+        assert force_const > 0.0
+        assert theta >= 0.0
+        assert theta <= np.pi
+        canonical_angles[dummy.canonicalize_bond(idxs)] = params
+
+    jkl = dummy.canonicalize_bond((j, k, l))
+
+    if jkl not in canonical_angles:
+        return False
+    elif canonical_angles[jkl][0] < 50.0:
+        return False
+    elif abs(canonical_angles[jkl][1] - 3.1415) < 0.05:
+        return False
+    elif abs(canonical_angles[jkl][1] - 0) < 0.05:
+        assert 0, "angle term of zero rad found."
+
+    return True
+
+
+def check_bond_angle_stability(j, k, l, bond_idxs, bond_params, angle_idxs, angle_params):
+    jk_stable = check_bond_stability(j, k, bond_idxs, bond_params)
+    kl_stable = check_bond_stability(k, l, bond_idxs, bond_params)
+    jkl_stable = check_angle_stability(j, k, l, angle_idxs, angle_params)
+    return jk_stable and kl_stable and jkl_stable
+
+
+def find_attached_dummy_atoms(dg, bond_idxs, anchor):
+    attached_dummy_atoms = []
+    for a, b in bond_idxs:
+        if a == anchor and b in dg:
+            attached_dummy_atoms.append(b)
+        elif b == anchor and a in dg:
+            attached_dummy_atoms.append(a)
+    return attached_dummy_atoms
+
+
+def find_junction_bonds(anchor, bond_idxs):
+    jbs = set()
+    for i, j in bond_idxs:
+        assert i != j
+        if i == anchor or j == anchor:
+            jbs.add(dummy.canonicalize_bond((i, j)))
+    return jbs
+
+
+def enumerate_anchor_groups(anchor_idx, bond_idxs, core_idxs):
+
+    # enumerate all 1 and 2 neighbor anchor atoms to form valid anchor groups.
+
+    assert anchor_idx in core_idxs
+    assert anchor_idx in [x[0] for x in bond_idxs] or anchor_idx in [x[1] for x in bond_idxs]
```

Comment:
nit: maybe slightly cleaner
```suggestion
    assert any(anchor_idx == atom_idx for bond in bond_idxs for atom_idx in bond)
```

Reply:
maybe

Reply:
minor, no need to change!